src/dtx/dtx_resync.c

```diff
@@ -0,0 +1,384 @@
+/**
+ * (C) Copyright 2019 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+/**
+ * dtx: resync DTX status
+ */
+#define D_LOGFAC	DD_FAC(dtx)
+
+#include <daos/placement.h>
+#include <daos/pool_map.h>
+#include <daos_srv/vos.h>
+#include <daos_srv/pool.h>
+#include <daos_srv/dtx_srv.h>
+#include <daos_srv/daos_server.h>
+#include <daos_srv/container.h>
+#include <abt.h>
+#include "dtx_internal.h"
+
+struct dtx_resync_entry {
+	d_list_t		dre_link;
+	struct daos_tx_entry	dre_dte;
+	uint64_t		dre_hash;
+	uint32_t		dre_intent;
+};
+
+#define dre_oid		dre_dte.dte_oid
+#define dre_xid		dre_dte.dte_xid
+
+struct dtx_resync_head {
+	d_list_t		drh_list;
+	int			drh_count;
+};
+
+struct dtx_resync_args {
+	struct ds_cont		*cont;
+	uuid_t			 po_uuid;
+	struct dtx_resync_head	 tables;
+	uint32_t		 version;
+};
+
+static inline void
+dtx_dre_release(struct dtx_resync_head *drh, struct dtx_resync_entry *dre)
+{
+	drh->drh_count--;
+	d_list_del(&dre->dre_link);
+	D_FREE_PTR(dre);
+}
+
+static int
+dtx_resync_commit(uuid_t po_uuid, struct ds_cont *cont,
+		  struct dtx_resync_head *drh, int count, uint32_t version,
+		  bool block)
+{
+	struct dtx_resync_entry		*dre;
+	int				 rc = 0;
+	int				 i = 0;
+
+	D_ASSERT(drh->drh_count >= count);
+
+	/* If we are in block mode, then we need to commit the DTXs as fast as
+	 * possible, so just add them into the CoS cache, then the general DTX
+	 * commit mechanism will commit them some time later.
+	 */
+	if (block) {
+		int	err;
+
+		do {
+			dre = d_list_entry(drh->drh_list.next,
+					   struct dtx_resync_entry, dre_link);
+			if (dre->dre_hash == 0)
```

Comment:
This is extremely unlikely, but hash can be zero... so you need a flag to indicate it's invalid instead of checking zero

Reply:
Zero hash is valid case for either punch object or the hash is just zero. We will commit such DTX synchronously without caching it.